src/routes/integrations/slack.ts

```diff
@@ -0,0 +1,257 @@
+      }
+
+      const existingIntegration = await con
+        .getRepository(UserIntegrationSlack)
+        .createQueryBuilder()
+        .where('"userId" = :userId', { userId: req.userId })
+        .andWhere(`meta->>'teamId' = :teamId`, { teamId: result.team.id })
+        .getOne();
+
+      const integrationMeta: IntegrationMetaSlack = {
+        appId: result.app_id,
+        slackUserId: result.authed_user.id,
+        scope: result.scope,
+        tokenType: result.token_type,
+        accessToken: await encrypt(
+          result.access_token,
+          process.env.SLACK_DB_KEY,
+        ),
+        teamId: result.team.id,
+        teamName: result.team.name,
+      };
+
+      if (existingIntegration) {
+        await con
+          .getRepository(UserIntegrationSlack)
+          .update({ id: existingIntegration.id }, { meta: integrationMeta });
+      } else {
+        await con.getRepository(UserIntegrationSlack).insert({
+          userId: req.userId,
+          meta: integrationMeta,
+        });
+      }
+
+      return redirectResponse({ res, path: redirectPath });
+    } catch (error) {
+      const isRedirectError = error instanceof RedirectError;
+
+      if (!isRedirectError) {
+        logger.error({ err: error }, 'error processing slack auth callback');
+      }
+
+      redirectResponse({
+        res,
+        path: redirectPath,
+        error: isRedirectError ? error : new RedirectError('internal error'),
+      });
+    }
+  });
+
+  fastify.post<{
+    Body: {
+      token: string;
+      challenge: string;
+      type: string;
+      team_id: string;
+      event: {
+        type: SlackEvent;
+      };
+    };
+    Headers: {
+      'x-slack-request-timestamp': string;
+      'x-slack-signature': string;
+    };
+  }>('/events', {
+    config: {
+      rawBody: true,
+    },
+    handler: async (req, res) => {
+      try {
+        if (req.body.type === SlackEventType.UrlVerification) {
+          const isValid = verifySlackSignature({ req });
+
+          return res.status(isValid ? 200 : 403).send({
+            challenge: isValid ? req.body.challenge : 'invalid signature',
+          });
+        }
+
+        const eventType = req.body.type;
+        const event = req.body.event?.type || 'unknown';
+
+        switch (event) {
+          case SlackEvent.AppUninstalled:
+          case SlackEvent.TokensRevoked: {
+            const teamId = req.body.team_id;
+
+            if (!teamId) {
+              logger.error(
+                {
+                  event,
+                  eventType,
+                },
+                'missing team id for slack event',
+              );
+
+              break;
+            }
+
+            const con = await createOrGetConnection();
+
+            await con
+              .getRepository(UserIntegrationSlack)
+              .createQueryBuilder()
+              .delete()
+              .where(`meta->>'teamId' = :teamId`, { teamId })
+              .execute();
```

Comment:
Basically, because we use bot token revoking it or removing app from workspace invalidates tokens for all users. So basically there will be some duplication of each `UserIntegration` row having same encrypted token but I think its ok since in the future this kind of table organization will serve us better if we have to obtain user token (which is specific for each user).